File: androidApp/src/main/java/com/walletline/android/presentation/screens/auth/verify_email/VerifyEmailViewModel.kt
package com.walletline.android.presentation.screens.auth.verify_email

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.walletline.android.presentation.screens.navArgs
import com.walletline.domain.model.OTPValidationMessage
import com.walletline.domain.use_case.auth.AuthUseCase
import com.walletline.domain.use_case.common.CommonUseCase
import com.walletline.domain.util.Resource
import com.walletline.presentation.screens.auth.verify_email.VerifyEmailState
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class VerifyEmailViewModel(
    savedStateHandle: SavedStateHandle,
    private val authUseCase: AuthUseCase,
    private val commonUseCase: CommonUseCase,
) : ViewModel(), VerifyEmailContract {

    private val _state = MutableStateFlow(VerifyEmailState())
    override val state: StateFlow<VerifyEmailState> = _state.asStateFlow()

    private val effectChannel = Channel<VerifyEmailContract.Effect>()
    override val effect: Flow<VerifyEmailContract.Effect> = effectChannel.receiveAsFlow()
    
    init {
        savedStateHandle.navArgs<VerifyEmailNavArg>().let { args ->
            _state.update { state -> state.copy(email = args.email) }
        }

        startTimer()
    }

    override fun onEvent(event: VerifyEmailContract.Event) {
        when (event) {
            is VerifyEmailContract.Event.OTPChange -> _state.update { state -> state.copy(otp = event.text) }
            VerifyEmailContract.Event.OnChangeEmailClicked -> effectChannel.trySend(
                VerifyEmailContract.Effect.NavigateToEmail
            )
            VerifyEmailContract.Event.OnContinueClicked -> verifyOTP()
            VerifyEmailContract.Event.OnPolicyClicked -> effectChannel.trySend(VerifyEmailContract.Effect.ShowPolicy)
            VerifyEmailContract.Event.OnResendClicked -> resendOTP()
            VerifyEmailContract.Event.OnTermsClicked -> effectChannel.trySend(VerifyEmailContract.Effect.ShowTerms)
        }
    }

    private fun startTimer() {
        val countDownTimer = commonUseCase.countDownTimer
        countDownTimer.start()
        countDownTimer.timer
            ?.onEach { tick ->
                _state.update { state ->
                    state.copy(timer = tick, isTimerActive = true)
                }
            }
            ?.onCompletion {
                _state.update { state ->
                    state.copy(isTimerActive = false)
                }
            }
            ?.launchIn(viewModelScope)
    }

    private fun verifyOTP() {
        _state.update { state -> state.copy(isLoading = true) }
        viewModelScope.launch {
            authUseCase.verifyOtp.execute(otp = state.value.otp).let { resource ->
                _state.update { state -> state.copy(isLoading = false) }
                when (resource) {
                    is Resource.Error -> handleVerifyError(resource.message)
                    is Resource.Success -> effectChannel.trySend(VerifyEmailContract.Effect.VerifySuccessful)
                }
            }
        }
    }

    private fun handleVerifyError(message: String?) {
        when {
            message?.contains("otp", ignoreCase = true) == true -> _state.update { it.copy(otpError = OTPValidationMessage.NotValid) }
            else -> effectChannel.trySend(VerifyEmailContract.Effect.ShowError(message ?: "Try again"))
        }
    }

    private fun resendOTP() {
        _state.update { state -> state.copy(isLoading = true) }
        viewModelScope.launch {
            _state.update { state -> state.copy(isLoading = false) }
            authUseCase.resendOtp.execute().let { resource ->
                when (resource) {
                    is Resource.Error -> effectChannel.send(VerifyEmailContract.Effect.ShowError(message = resource.message ?: "Try again"))
                    is Resource.Success -> effectChannel.send(VerifyEmailContract.Effect.ResendOtpSuccess(resource.data.otp)).also { startTimer() }
                }
            }
        }
    }

}

File: iosApp/iosApp/presentation/screens/email_verify/VerifyEmailViewModel.swift
//
//  EmailVerifyViewModel.swift
//  iosApp
//
//  Created by Mostafa Hosseini on 5/7/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Combine
import Foundation
import KMPNativeCoroutinesCombine
import shared

@MainActor
class VerifyEmailViewModel: ObservableObject {
    private let useCase = KoinHelper()
    private var cancellables = Set<AnyCancellable>()

    @Published private(set) var state = VerifyEmailState()
    private let eventSubject = PassthroughSubject<VerifyEmailEffect, Never>()

    func effect() -> AnyPublisher<VerifyEmailEffect, Never> {
        return eventSubject.eraseToAnyPublisher()
    }

    init(cancellables: Set<AnyCancellable> = Set<AnyCancellable>(), state: VerifyEmailState = VerifyEmailState()) {
        self.cancellables = cancellables
        self.state = state
        startTimer()
    }

    func onEvent(_ event: VerifyEmailEvent) {
        switch event {
        case .OTPChange(let text):
            updateState(otp: text)
        case .OnChangeEmailClicked:
            eventSubject.send(.NavigateToEmail)
        case .OnContinueClicked:
            verifyOTP()
        case .OnResendClicked:
            resendOTP()
        case .OnPrivacyClicked:
            doNothing()
        case .OnTermsClicked:
            doNothing()
        case .cancelAsyncTask:
            cancelAsyncTasks()
        case .UpdateEmail(let email):
            updateState(email: email)
        }
    }

    private func verifyOTP() {
        let verifyFuture = createFuture(for: useCase.verifyOtp(otp: state.otp))
        updateState(isLoading: true)
        verifyFuture.sink(
            receiveCompletion: { _ in },
            receiveValue: { [self] res in
                updateState(isLoading: false)
                switch res {
                case is ResourceError<KotlinNothing>:
                    handleVerifyError((res as? ResourceError<KotlinNothing>)?.message)
                case is ResourceSuccess<KotlinBoolean>:
                    eventSubject.send(.VerifySuccessful)
                default:
                    doNothing()
                }
            }
        )
        .store(in: &cancellables)
    }

    private func resendOTP() {
        let resendOTPFuture = createFuture(for: useCase.resendOtp())
        updateState(isLoading: true)
        resendOTPFuture.sink(
            receiveCompletion: { _ in },
            receiveValue: { [self] res in
                updateState(isLoading: false)
                switch res {
                case is ResourceError<KotlinNothing>:
                    handleVerifyError((res as? ResourceError<KotlinNothing>)?.message)
                case is ResourceSuccess<OtpData>:
                    startTimer()
                    guard let res = res as? ResourceSuccess<OtpData> else { return }
                    guard let otp = res.data?.otp else { return }
                    eventSubject.send(.ResendOTP(otp: otp))
                default:
                    doNothing()
                }
            }
        )
        .store(in: &cancellables)
    }

    private func handleVerifyError(_ errorMessage: String?) {
        if errorMessage?.contains("otp") == true {
            updateState(otpError: OTPValidationMessageNotValid())
        } else {
            eventSubject.send(.Error(message: errorMessage ?? "Try again"))
        }
    }

    private func startTimer() {
        useCase.startTimer()
        guard let timer = useCase.timer() else { eventSubject.send(.Error(message: "timer null"));return }
        let publisher = createPublisher(for: timer)
        
        publisher.sink(receiveCompletion: { _ in
            self.updateState(isTimerActive: false)
        }, receiveValue: { value in
            self.updateState(timer: Int32(truncating: value))
        })
        .store(in: &cancellables)
    }

    private func updateState(
        otp: String? = nil,
        otpError: OTPValidationMessage? = nil,
        email: String? = nil,
        timer: Int32? = nil,
        isTimerActive: Bool? = nil,
        isLoading: Bool? = nil,
        isActionsEnabled: Bool? = nil
    ) {
        guard let currentState = state.copy() as? VerifyEmailState else { return }
        DispatchQueue.main.async { [self] in
            state = state.doCopy(
                otp: otp ?? currentState.otp,
                otpError: otpError ?? currentState.otpError,
                email: email ?? currentState.email,
                timer: timer ?? currentState.timer,
                isTimerActive: isTimerActive ?? currentState.isTimerActive,
                isLoading: isLoading ?? currentState.isLoading,
                isActionsEnabled: isActionsEnabled ?? currentState.isActionsEnabled
            )
        }
    }

    private func cancelAsyncTasks() {
        cancellables.forEach { cancellable in
            cancellable.cancel()
        }
    }

    private func doNothing() {}

    @inline(__always)
    private func handleAction(block: () -> Void) {
        updateState(isActionsEnabled: false)
        block()
        updateState(isActionsEnabled: true)
    }
}

enum VerifyEmailEvent {
    case OTPChange(text: String)
    case UpdateEmail(email: String)
    case OnContinueClicked
    case OnChangeEmailClicked
    case OnResendClicked
    case OnTermsClicked
    case OnPrivacyClicked
    case cancelAsyncTask
}

enum VerifyEmailEffect {
    case Error(message: String)
    case VerifySuccessful
    case NavigateToEmail
    case NavigateToPattern
    case ShowPolicy
    case ShowTerms
    case ResendOTP(otp: String)
}
